### PR TITLE
INFERRED_NO_EXCLUSION_LIST should take year into account

### PR DESCRIPTION
**Pull request by bethanyaconnor**
A month ago, in #46819, I added schools to this list I believe for the first time ever. What we wanted was to account for the fact that certain states, like Michigan, only gave us data on schools where they went from not teaching CS to teaching CS instead of a list of all schools that teach CS, like in past years. For schools that aren't in the new data sets, we want to infer `teaches_cs` from the previous year's data.

However, this caused an unintended downstream effect where previous years, when these states provided more complete data, were also affected because this list applied for a state for *all years*. The fix here is to make this list depend on the year as well as the state.

An example of this issue:

Correct case:
School X in Michigan does not teach CS and School Y does. Last year, Michigan gave us a complete list of schools that taught CS, so we know School Y taught CS and we should be able to infer School X did not teach CS last year. This year, Michigan only gave us a list of schools where `teaches_cs` changed, so School X should be noted as `HN` (historic no) and School Y should be noted as `HY` (historic yes).

Current case:
School X in Michigan does not teach CS and School Y does. Last year, Michigan gave us a complete list of schools that taught CS, so we know School Y taught CS and we should be able to infer School X did not teach CS last year. This worked *last year* but due to #46819, School X's status has become either `NULL` or `MAYBE`. That means that, this year, School Y is correctly noted as `HY` but School X is still a `NULL` or `MAYBE`.

Because we essentially lose the `N` and `HN` schools, our data is all wrong. I believe that tying the inclusion list to the year will resolve this particular issue.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by maureensturgeon — APPROVED** [review]

**Comment by bencodeorg** [comment]
> A month ago, in https://github.com/code-dot-org/code-dot-org/pull/46819, I added schools to this list I believe for the first time ever.

Just noting the `INFERRED_NO_EXCLUSION_LIST` has been populated before, eg at this point in history:

https://github.com/code-dot-org/code-dot-org/commit/2d76e3c4f2820e319ded7b1a443baf5ac3ae8686#

**Comment by bencodeorg** [comment]
> only gave us data on schools where whether or not they teach CS has changed instead of a list of all schools that teach CS, like in past years.

Nit: my understanding now is that the list is comprised of schools that are newly teaching CS, not all changes (eg, if they went from teaching -> not teaching), correct?

File: dashboard/test/models/census/census_summary_test.rb [hunk 1]
```diff
@@ -262,21 +262,38 @@ class Census::CensusSummaryTest < ActiveSupport::TestCase
...
       :with_one_year_ago_teaches_yes,
       :with_two_years_ago_teaches_yes,
       :with_three_years_ago_teaches_yes,
-      state: Census::StateCsOffering::INFERRED_NO_EXCLUSION_LIST.first,
+      state: state,
       school_year: school_year
     validate_summary(school, school_year, "MAYBE")
   end
 
+  test "High school lack of state data in a denylist state and non-denylist year does not override anything" do
```
Comment (bencodeorg) [inline dashboard/test/models/census/census_summary_test.rb]: Know you're in a hurry so can leave this comment for someone else to deal with, but I think this test is showing that we **do override** when we're in a state-year that is not in the deny list?

**Review by bencodeorg — APPROVED** [review]
Makes sense to me!

As discussed offline, when the dust has settled, I do think it's worth confirming whether this 2021-2022 data should represent historical or current yes's for the many schools that are in the 2020-2021 data, but not in 2021-2022. If they should be current yes's, appending the new 2021-2022 data to the bottom of the 2020-2021 data and calling that combination the 2021-2022 would make sense to me. Maybe @Sam-Hutchins can comment.

File: dashboard/test/models/census/census_summary_test.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/test/models/census/census_summary_test.rb]: omg yes you're right. I can update in this PR but maybe not wait for drone -- that's a pretty big error

**Comment by bethanyaconnor** [comment]
> Nit: my understanding now is that the list is comprised of schools that are newly teaching CS, not all changes (eg, if they went from teaching -> not teaching), correct?

Yes that is my understanding as well. It sounds like we're okay assuming that no school stopped teaching CS for these schools this year.

**Comment by bencodeorg** [comment]
> Yes that is my understanding as well. It sounds like we're okay assuming that no school stopped teaching CS for these schools this year.

Sounds good, I'd just suggest an update to the description to make it clear the new data represents not teaching -> teaching, not all changes.

**Comment by bethanyaconnor** [comment]
> > A month ago, in https://github.com/code-dot-org/code-dot-org/pull/46819, I added schools to this list I believe for the first time ever.

> Just noting the INFERRED_NO_EXCLUSION_LIST has been populated before, eg at this point in history:


Oops yeah I noticed this later. I think it is the first time states have been re-added to that list though

**Comment by bethanyaconnor** [comment]
Only change is a test name change that I ran locally. Merging now

**Pull request merged**